File: leetcode/java/0013. Roman to Integer.java
package leetcode.java;

import java.util.HashMap;

// Time complexity: O(s), where s is the length of the string
// Space complexity: O(1)

class Solution {
    public int romanToInt(String s) {
        HashMap<Character, Integer> symsToVals = initializeMeanings();
        int res = 0;
        
        for (int i=0; i<s.length(); i++) {

            if (i < s.length() - 1 && s.charAt(i) == 'I' && (s.charAt(i+1) == 'V' || s.charAt(i+1) == 'X')) {
                res -= symsToVals.get(s.charAt(i));
            } else if ( i < s.length() - 1 && s.charAt(i) == 'X' && (s.charAt(i+1) == 'L' || s.charAt(i+1) == 'C')) {
                res -= symsToVals.get(s.charAt(i));
            } else if (i < s.length() - 1 && s.charAt(i) == 'C' && (s.charAt(i+1) == 'D' || s.charAt(i+1) == 'M')) {
                res -= symsToVals.get(s.charAt(i));
            } else {
                res += symsToVals.get(s.charAt(i));
            }
        }
        return res;
    }

    public HashMap<Character, Integer> initializeMeanings() {
        char[] symbols = new char[]{'I', 'V', 'X', 'L', 'C', 'D', 'M'};
        int[] values = new int[]{1, 5, 10, 50, 100, 500, 1000};

        HashMap<Character, Integer> res = new HashMap<>();

        for (int i=0; i<symbols.length; i++) {
            res.put(symbols[i], values[i]);
        }
        
        return res;
    }
}

File: leetcode/python/0013. Roman to Integer.py
# Time complexity: O(s), where s is the length of the string
# Space complexity: O(1)

class Solution(object):
    def romanToInt(self, s):
        res = 0
        skip = False
        
        mapping = {
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000,
        }

        for i in range(len(s)):
            if skip:
                skip = not skip
                continue
            if i < len(s) - 1 and s[i] == "I" and (s[i+1] == "V" or s[i+1] == "X"):
                if s[i+1] == "V":
                    res += 4
                else:
                    res += 9
                skip = True
            elif i < len(s) - 1 and s[i] == "X" and (s[i+1] == "L" or s[i+1] == "C"):
                if s[i+1] == "L":
                    res += 40
                else:
                    res += 90
                skip = True
            elif i < len(s) - 1 and s[i] == "C" and (s[i+1] == "D" or s[i+1] == "M"):
                if s[i+1] == "D":
                    res += 400
                else:
                    res += 900
                skip = True
            else:
                res += mapping[s[i]]
        return res